libraries/chain/controller.cpp

```diff
@@ -1274,6 +1274,57 @@ struct controller_impl {
       apply_handlers[receiver][make_pair(contract,action)] = v;
    }
 
+   void set_trx_expiration(signed_transaction& trx) {
+      if (is_builtin_activated(builtin_protocol_feature_t::no_duplicate_deferred_id)) {
+         trx.expiration       = time_point_sec();
+         trx.ref_block_num    = 0;
+         trx.ref_block_prefix = 0;
+      } else {
+         trx.expiration = time_point_sec{
+            pending_block_time() + fc::microseconds(999'999)}; // Round up to nearest second to avoid appearing expired
+         trx.set_reference_block(chain_head.id());
+      }
+   }
+
+   getpeerkeys_res_t get_top_producer_keys(fc::time_point deadline) {
+      try {
+         auto get_getpeerkeys_transaction = [&]() {
+            auto perms = vector<permission_level>{};
+            action act(perms, config::system_account_name, "getpeerkeys"_n, {});
+            signed_transaction trx;
+
+            trx.actions.emplace_back(std::move(act));
+            trx.set_reference_block(chain_head.id());
+            set_trx_expiration(trx);
+            return trx;
+         };
+
+         auto metadata = transaction_metadata::create_no_recover_keys(
+            std::make_shared<packed_transaction>(get_getpeerkeys_transaction()),
+            transaction_metadata::trx_type::read_only);
+
+         auto trace = push_transaction(metadata, deadline, fc::microseconds::maximum(), 0, false, 0);
```

Comment:
I think we should limit this read-only trx runtime to something reasonable. How about, 10ms.

Reply:
I did `20ms` because I am a little bit concerned that it does a number of db accesses and 10ms might be short in our overwhelmed ci/cd, but if you think it is too much please let me know and I'll change it.

Reply:
https://github.com/AntelopeIO/spring/pull/1316/commits/ca9c2f193a449e760e44c5c0f9f7cf1bcde1eb65